### ISSUE TITLE
Evaluation as a stand-alone job over data, not the trigger for multiple jobs

### ISSUE THREAD

**Issue by ividal** [mozilla-ai/lumigator#435]
**Is your feature request related to a problem? Please describe.**
So far, Lumigator's evaluation was actually a combination of these tasks/jobs:
* predictions from a model under test --> "predictions"
* [optionally, if true human-sourced ground-truth was missing] predictions from a reference model  --> "annotations"
* comparison of predictions and annotations to obtain a quantitative measure of how close the model under test performed w.r.t. to either the human annotator or the reference model.

This meant launching at least 1 inference job to get predictions per evaluation. This started to change with #328. This issue would wrap that up by having evaluation only ingest data (the actual annotations+predictions), rather than triggering jobs to generate the data.

**Describe the solution you'd like**
An evaluation job should only require data (a prediction from a model under test and some annotations, regardless of what the source was).

**Describe alternatives you've considered**
Part of the discussion is whether to include here a check that annotations and predictions did not come from the same model. Should this be done at experiment level?

**Additional context**
Draft work to collect feedback can be found in #333 .

**Comment by ividal** [comment]
The scope of this is as discussed originally with @veekaybee, prior to any decisions in #437.

Should something need to change after that, it will be captured [here](https://github.com/orgs/mozilla-ai/projects/6/views/1?pane=issue&itemId=89419116).

**Comment by veekaybee** [comment]
Per our discussion, acknowledging that for the scope of this I'll be:

1. Picking up the draft PR I had created in https://github.com/mozilla-ai/lumigator/pull/333 - I'll likely be closing out this PR since the codebase has shifted significantly since then
2. Splitting out evaluation into its own job
3. This work is independent of other ongoing work on evaluation.
4. @peteski22 and @javiermtorres will be buddies on this task, reviewing a draft of the PR.

**Comment by veekaybee** [comment]
As a heads-up in the last iteration of this work, we ended up creating an eval-lite which shadowed evaluation in anticipation of breaking APIs for upcoming demos. Given that inference is now fully built out and we have more time to test, I'll be replacing evaluation entirely instead of duplicating flows to reduce code duplication and errors.

**Comment by veekaybee** [comment]
Adding information about inputs and outputs of job:

![image](https://github.com/user-attachments/assets/b6f6e1da-7feb-492e-852b-f597ada00603)

**Comment by ividal** [comment]
Thanks for this start @veekaybee 
Something to start the discussion and also help start designing a workflow:

* What is the schema for the input data? E.g. does this job expect a single file with predictions and annotations or two files of the same type, simply one under "ground-truth" and another under predictions, or something else entirely?
* What does the config schema look like / what (if any) fields are shared with the inference job that was carved out?
* Other than data and config schemas, any other big changes w.r.t. the combined job due to separating these two jobs?

**Comment by aittalam** [comment]
TY @veekaybee! I agree on starting with what the current inference job provides first, so we make sure the two jobs can be correctly piped, then creating a new issue where we consolidate the approach to saving / loading data within the system. WDYT about taking advantage of this work to also introduce the possibility for users to use custom field names, i.e. instead of providing a simple dataset we'd have possibly multiple (dataset, field_name) pairs, vs having a job that does this for us? I generally like the idea and I think even if we don't do it in this first iteration we can at least create a specific issue for that but I am happy to hear your thoughts abt it (also if you think this might make things easier for you I'd say just go ahead with it!)

A heads up abt potential headaches (not all strictly related to the JSON choice):

- the current evaluator code uses [datasets.load_from_disk](https://huggingface.co/docs/datasets/v3.2.0/en/package_reference/loading_methods#datasets.load_from_disk) to load a dataset saved with the HF datasets lib on S3. Loading from S3 is automatic if you import `s3fs`, so you don't have to worry about copying stuff locally. My understanding is that if you want to read files in other formats you'll have to use [datasets.load_dataset](https://huggingface.co/docs/datasets/v3.2.0/en/package_reference/loading_methods#datasets.load_dataset) instead and that does not work with S3 out of the box (see [here](https://discuss.huggingface.co/t/how-to-use-s3-path-with-load-dataset-with-streaming-true/25739), [here](https://github.com/huggingface/datasets/issues/5281))

- the way things work now, as we want eval to run from ray jobs, we have to store inference output as a lumigator dataset (no matter whether it is CSV, JSON, or HF datasets format, but it has to be on our object store) and then pass the id to evaluation. Is this still gonna be the case? If so, I think it'd be heplful for you to use some inference results pre-uploaded in the format you expect while someone works in parallel on a PR that aligns inference to it. WDYT? If that's ok for you, let's agree on the format you'd expect and I am happy to create an issue for generating data which is compatible with it.

**Comment by veekaybee** [comment]
I think we can potentially include 1, do we see it being passed in the config here? 

```python

class DatasetConfig(BaseModel):
    path: str
    fields: dict # this is the one to add, with a default of prediction and ground_truth, and then we'd want to change the constraint on the dataset side  as well


class EvaluationJobConfig(BaseModel):
    metrics: list[str] = Field(default=["rouge", "meteor", "bertscore"])
    use_pipeline: bool = False
    max_samples: int = 0
    return_input_data: bool = False
    return_predictions: bool = False
    storage_path: str = ""


class TokenizerConfig(BaseModel):
    path: str


class EvalJobConfig(BaseModel):
    name: str | None
    dataset: DatasetConfig
    model: str | None = None
    tokenizer: TokenizerConfig
    evaluation: EvaluationJobConfig
    model_config = ConfigDict(extra="forbid")
```

**Comment by aittalam** [comment]
> We'd talked about simplifying or excluding HF datasets eventually in favor of json in the app. It's something that makes a lot of sense for the future as the data we are passing around is not that large. I'd like to explore in this PR what that might look like given the diagram above if it works for you to create json files. I think it would "just" be converting dialogsum to json 

Yeah, a json dialogsum would be a good start! My suggestion then would be to manually store it on S3 and try calling your job by directly passing it the S3 path to make sure it works properly.

If you want to run it in lumigator (i.e. as a ray job submission) from the very start this might make things a bit more complex as you'd need to find a way to make the dataset available in the DB too. If you want to go this way, I'd suggest to manually create a DB entry for testing. 

The reason is that RN we [just allow for CSV files](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/backend/backend/services/datasets.py#L63) as datasets and the work required to deal with json datasets might involve:

- adding proper CSV / JSON dataset formats in [schemas](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/schemas/lumigator_schemas/datasets.py) (at last!)
- updating [validate_dataset_format](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/backend/backend/services/datasets.py#L46) accordingly
- updating [validate_experiment_dataset](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/backend/backend/services/datasets.py#L63) so it is able to open a JSON file too
- updating [_save_dataset_to_s3](https://github.com/mozilla-ai/lumigator/blob/main/lumigator/python/mzai/backend/backend/services/datasets.py#L113) so it allows for saving json files differently
- uploading the json dataset so you can use it in your own experiment

**Comment by aittalam** [comment]
> I think we can potentially include 1, do we see it being passed in the config here?
> 
> class DatasetConfig(BaseModel):
>     path: str
>     fields: dict # this is the one to add, with a default of prediction and ground_truth, and then we'd want to change the constraint on the dataset side  as well

Yes, I think that'd work!

**Comment by veekaybee** [comment]
@ividal Are there any additional points of discussion or clarification you'd like to see for this issue?

**Comment by veekaybee** [comment]
> Yeah, a json dialogsum would be a good start! My suggestion then would be to manually store it on S3 and try calling your job by directly passing it the S3 path to make sure it works properly.
> 
> If you want to run it in lumigator (i.e. as a ray job submission) from the very start this might make things a bit more complex as you'd need to find a way to make the dataset available in the DB too. If you want to go this way, I'd suggest to manually create a DB entry for testing.

For this piece, my approach would be to start with the simplest small thing that works, aka manual storage, since we are already changing so many other moving parts, but acknowleding that we want to move towards JSON throughout the application as a transport format - does this sound ok and are we ok committing to that for the future?

**Comment by ividal** [comment]
> WDYT about taking advantage of this work to also introduce the possibility for users to use custom field names, 

Caution here (small PRs, etc.), specially since this is a refactoring: please start with a minimal change necessary, add new functionality in a later PR.

Please exclude exploring formats, custom fields... That all can be done in the background while this first piece is going in and workflows are built.

I agree on adding other more sane formats than csv sooner rather than later (I don't think _removing_ hf is necessary... but maybe I´m missing context). Either way, I wouldn't want to mix that discussion with this particular piece.

This way, the moment this PR goes in, an initial workflow in #479 could ideally go live too, which is what we need first. Removing/adding formats, customizable fields, etc. are great material for follow-up PRs.

Hope this makes sense!

**Comment by veekaybee** [comment]
> Please exclude exploring formats, custom fields... That all can be done in the background while this first piece is going in and workflows are built.

In that case, what I'm understanding is that I'm proceeding only with everything here, no JSON, and no changes in the field names, can we confirm? 

https://github.com/mozilla-ai/lumigator/issues/435#issuecomment-2532571026

**Comment by ividal** [comment]
Sure, I agree with [that](https://github.com/mozilla-ai/lumigator/issues/435#issuecomment-2532571026)!

**Comment by ividal** [comment]
Sure, I agree with that with [that](https://github.com/mozilla-ai/lumigator/issues/435#issuecomment-2532571026, except no json, no change in names. Thanks for the discussion! 🙏

**Comment by veekaybee** [comment]
Great, confirming that I'm starting a PR based on the above. Thanks!

**Comment by veekaybee** [comment]
Work in progress https://github.com/mozilla-ai/lumigator/pull/499
